src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+        const surfaceArea = extent.x * extent.y + extent.y * extent.z + extent.z * extent.x;
+
+        return node.triCount * surfaceArea;
+    }
+
+    /**
+     * Subdivides the BVH on a node
+     *
+     * @param {number} nodeIdx - The index of the node to subdivide at
+     * @ignore
+     */
+    Subdivide(nodeIdx) {
+        const node = this.bvhNode[nodeIdx];
+
+        const { triangles, triIdx } = this;
+
+        const splitDetails = { node: node, axis: null, splitPos: null, splitCost: null };
+
+        // const splitCost = this.findBestSplitPlane(splitDetails);
+        this.findBestSplitPlane(splitDetails);
+
+        const axis = splitDetails.axis;
+
+        const splitPos = splitDetails.splitPos;
+
+        // const noSplitCost = this.calculateNodeCost(node);
+
+        // terminates tree building based on split cost
+        // if (splitCost >= noSplitCost) {
+        //     return;
+        // }
+
+        // terminates tree building when number of leafnodes is less than 5
+        if (node.triCount < 10) {
+            return;
+        }
+
+        // Perform the split
+        let i = node.leftFirst;
+        let j = i + node.triCount - 1;
+        while (i <= j) {
+            if (triangles[triIdx[i]].centroid[axis] < splitPos) {
+                i++;
+            } else {
+                [triIdx[i], triIdx[j]] = [triIdx[j], triIdx[i]];
+                j--;
+            }
+        }
+
+        const leftCount = i - node.leftFirst;
+        if (leftCount === 0 || leftCount === node.triCount) {
+            return;
+        }
+
+        // Create child nodes for each half
+        const leftChildIdx = this.nodesUsed++;
+        const rightChildIdx = this.nodesUsed++;
+        this.bvhNode[leftChildIdx].leftFirst = node.leftFirst;
+        this.bvhNode[leftChildIdx].triCount = leftCount;
+        this.bvhNode[rightChildIdx].leftFirst = i;
+        this.bvhNode[rightChildIdx].triCount = node.triCount - leftCount;
+        node.leftFirst = leftChildIdx;
+        node.triCount = 0;
+        this.updateNodeBounds(leftChildIdx);
+        this.updateNodeBounds(rightChildIdx);
+
+        // Recurse into each of the child nodes
+        this.Subdivide(leftChildIdx);
+        this.Subdivide(rightChildIdx);
+    }
+
+    /**
+     * Builds the BVH
+     *
+     * @ignore
+     */
+    buildBVH() {
+        const { triangles } = this;
+        const N = triangles.length;
+        this.triIdx = Array.from({ length: N }, (x, i) => i);
+        this.bvhNode = Array.apply(null, Array(N * 2)).map(function () {
+            return new BVHNode();
+        });
+        const rootNodeIdx = 0;
+        this.nodesUsed = 1;
+
+        for (let i = 0; i < N; i++) {
+            triangles[i].centroid = new Vec3();
+            triangles[i].centroid.add(triangles[i].vertex0);
+            triangles[i].centroid.add(triangles[i].vertex1);
+            triangles[i].centroid.add(triangles[i].vertex2);
+            triangles[i].centroid.mulScalar(1 / 3);
```

Comment:
this should be a function on a triangle. And perhaps use divScalar(3); (which should be changed to do a single division only, and multiply with it)

Reply:
Declared a constant 1/3 at the top of the file which is used in mulScalar.